# Credit Risk Classification

Credit risk poses a classification problem that’s inherently imbalanced. This is because healthy loans easily outnumber risky loans. In this Challenge, you’ll use various techniques to train and evaluate models with imbalanced classes. You’ll use a dataset of historical lending activity from a peer-to-peer lending services company to build a model that can identify the creditworthiness of borrowers.

## Instructions:

This challenge consists of the following subsections:

* Split the Data into Training and Testing Sets

* Create a Logistic Regression Model with the Original Data

* Predict a Logistic Regression Model with Resampled Training Data 

### Split the Data into Training and Testing Sets

Open the starter code notebook and then use it to complete the following steps.

1. Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

2. Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

    > **Note** A value of `0` in the “loan_status” column means that the loan is healthy. A value of `1` means that the loan has a high risk of defaulting.  

3. Check the balance of the labels variable (`y`) by using the `value_counts` function.

4. Split the data into training and testing datasets by using `train_test_split`.

### Create a Logistic Regression Model with the Original Data

Employ your knowledge of logistic regression to complete the following steps:

1. Fit a logistic regression model by using the training data (`X_train` and `y_train`).

2. Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

3. Evaluate the model’s performance by doing the following:

    * Calculate the accuracy score of the model.

    * Generate a confusion matrix.

    * Print the classification report.

4. Answer the following question: How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

### Predict a Logistic Regression Model with Resampled Training Data

Did you notice the small number of high-risk loan labels? Perhaps, a model that uses resampled data will perform better. You’ll thus resample the training data and then reevaluate the model. Specifically, you’ll use `RandomOverSampler`.

To do so, complete the following steps:

1. Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

2. Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

3. Evaluate the model’s performance by doing the following:

    * Calculate the accuracy score of the model.

    * Generate a confusion matrix.

    * Print the classification report.
    
4. Answer the following question: How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

### Write a Credit Risk Analysis Report

For this section, you’ll write a brief report that includes a summary and an analysis of the performance of both machine learning models that you used in this challenge. You should write this report as the `README.md` file included in your GitHub repository.

Structure your report by using the report template that `Starter_Code.zip` includes, and make sure that it contains the following:

1. An overview of the analysis: Explain the purpose of this analysis.


2. The results: Using bulleted lists, describe the balanced accuracy scores and the precision and recall scores of both machine learning models.

3. A summary: Summarize the results from the machine learning models. Compare the two versions of the dataset predictions. Include your recommendation for the model to use, if any, on the original vs. the resampled data. If you don’t recommend either model, justify your reasoning.

In [46]:
# Import the modules
# Here we import all the packages and dependencies including two new functions from the 
# sklearn module and one new function from imblearn module, along with the usual Path function
# pathlib and pandas as pd and numpy as np.
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from imblearn.metrics import classification_report_imbalanced

import warnings
warnings.filterwarnings('ignore')

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [70]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
# Here we use the pandas read_csv() function to read in the csv file labeled lending_data.csv. Pandas reads in the data and 
# turns the data into a dataframe.
credit_df = pd.read_csv(Path("./Resources/lending_data.csv"))

# Review the DataFrame
# Here we call the dataframe so we can review it.
credit_df

loan_size  interest_rate  borrower_income  debt_to_income  \
0        10700.0          7.672            52800        0.431818   
1         8400.0          6.692            43600        0.311927   
2         9000.0          6.963            46100        0.349241   
3        10700.0          7.664            52700        0.430740   
4        10800.0          7.698            53000        0.433962   
...          ...            ...              ...             ...   
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
0                    5                 1       22800            0  
1                    3                 0       13600            0  
2                    3                 0       16100            0  
3                    5                 1       22700            0  
4                    5                 1       23000            0  
...                ...               ...         ...          ...  
77531               12                 2       56600            1  
77532               11                 2       50900            1  
77533               11                 2       50300            1  
77534               10                 2       45300            1  
77535                9                 2       42300            1  

[77536 rows x 8 columns]

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [48]:
# Separate the data into labels and features

# Separate the y variable, the labels
# Here we are splitting up the data, in the y variable we assign the column labeled 'loan_status', this is what we are 
# going to try and predict with the logistic regression model. This is called the target.
y = credit_df['loan_status']

# Separate the X variable, the features
# Here we assign all the columns in the credit_df dataframe to X, less the 'loan_status' colmun, this is acheived with the 
# .drop() function. So we are saying assign all columns of credit_df but drop the 'loan_status'. This data is the information
# that we will ask the module to use to predict the target with. So thats why we give X all the columns of data, so it can
# apply the algorithims to the data and give us a prediction.
X = credit_df.drop(columns=['loan_status'])

In [82]:
# Review the y variable Series
# Here we call the variable 'y' to see whats in the series. This is the 'loan_status' column from our original dataframe.
# The .head() function returns the first 5 rows in the series.
y.head()

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [81]:
# Review the X variable DataFrame
# Here we review the X variable by calling it. We can see that indeed the loan_status' column has been removed. And
# the .head() function returns the first 5 rows in the series.
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [84]:
# Check the balance of our target values
# Its helpful to use the value_counts() function to see how the test & train data is weighted. The majority is 75036
# and the minority is 2500. This is an imbalanced class.
y.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [52]:
# Import the train_test_learn module
# We import the train_test_split function from the module sklearn.model_selection.
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
# Here we split the feature dataset into training and testing sets, then we can fit to a logistic regression model.
# The reason we split the data is so the algorithims have some data to learn how to predict the rest of the data that we 
# want it to predict. The algorithims need something to fit or train with, so we split the data.
# The random_state controls the shuffling of the data before it makes the split of data.
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [53]:
# Import the LogisticRegression module from SKLearn
# Here we import the LogisticRegression function from the sklearn.linear_model module.
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
# Here we are creating a Logistic Regression function and assign it to a variable named model.
model = LogisticRegression(random_state=1)

# Fit the model using training data
# Here we are fitting or training the model to the data from the original dataframe. Remember all the columns from the 
# original dataframe we assigned to X, this is the data that we pass to the model to train with. 
lr_model = model.fit(X_train, y_train)
lr_model

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [92]:
# Make a prediction using the testing data
# Here we use the .predict() function to predict with the data from the features of credit_df, and we assign it to a new
# variable named y_pred for use later. Note, this is where all the work comes together and the model actually makes the predictions
# for our target.
y_pred = lr_model.predict(X_test)
y_pred

array([0, 0, 0, ..., 0, 0, 0])

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [93]:
# Print the balanced_accuracy score of the model
# Here we use the balanced_accuracy_score() function wich calculates the accuracy of our model predictions for the
# testing data.
balanced_accuracy_score(y_test, y_pred)

0.9520479254722232

In [56]:
# Generate a confusion matrix for the model
# Here we use the confusion_matrix() function and pass as parameters the y_test and y_pred
# variables, we get a confusion matrix. This helps us to understand how many predicted true/
# actaully true = 18663, predicted false/actually true = 102, predicted false/actually true = 56,
# predicted false / actually false = 563. Basically the confusion matrix seperates into 2 columns
# all the data that was predicted true and false and wether it was indeed true or false.
confusion_matrix(y_test, y_pred)

array([[18663,   102],
       [   56,   563]])

In [57]:
# Print the classification report for the model
# Here we print the classification_report_imbalanced() function, and its helpfull to show us 
# several metrics, including presicion, recall, F1 and support. We pass both the test data and the 
# actual prediction data, just like we did for the confusion_matrix(). We are able to glean a lot of
# good data from the classification_matrix(), specifically presicion, recall and F1.
print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

          0       1.00      0.99      0.91      1.00      0.95      0.91     18765
          1       0.85      0.91      0.99      0.88      0.95      0.90       619

avg / total       0.99      0.99      0.91      0.99      0.95      0.91     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** With the sklearn module we are able to call upon the classification_report() function and get a pretty good idea of how the Logistic Regression algorithim predicted the test and predict data. Here we see the values that the model observed as 0 or healthy loans, 100% of the time the model predicted correctly, and all the values that the model observed as 1 high-risk loans, 85% of the time it predicted them correctly. Also 99% of the time the model predicted that the healthy loans would not default and 91% of the time the high risk loans would default. We know that the training dataset was imbalanced so maybe we can make some adjustments and get the recall and precision higher.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [58]:
# Import the RandomOverSampler module form imbalanced-learn
# Here we import the RandomOverSampler function from the imblearn.over_sampling module. 
# RandomOverSampler is a function that randomly samples from the minority dataset to increase
# the training dataset to more equal the majority dataset, in an attempt to deal with the imbalanced
# sets of data that we are training our model with vrs. the data that we are predicting with. So
# the RandomOverSampler over samples from the minority group untill the training dataset equals
# the majority.
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
# Here we instantiate or create a model for oversampling.
random_oversampler = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
# Here we fit or train the oversampler model to all the training data, y_train and X_train.
# This represents all the data we have designated as training data.
X_resampled, y_resampled = random_oversampler.fit_resample(X_train, y_train)

In [59]:
# Count the distinct values of the resampled labels data
# This is the value_counts() function and it shows us that since we used the RandomOverSamper 
# function and oversampled from the minority train dataset, we now have a equal train and test dataset.
# Note that we didn't take data from the majority and give it to the minority, instead we randomly over
# smpled the minority untill it equalled the majority. For the first model we used, the train dataset
# had 2500 values in it, and now that we RandomOverSampled we see that the train sataset has 56271 values
# in it.
y_resampled.value_counts()

0    56271
1    56271
Name: loan_status, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [63]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
# Here we instantiate or create a model to use the LogisticRegression algorithim. 
resample_model = LogisticRegression(random_state=1)

# Fit the model using the resampled training data
# And here we fit or train the LogisticRegression model to the X_resampled and y_resampled datasets.
lr_resampled_model = resample_model.fit(X_resampled, y_resampled)

# Make a prediction using the testing data
# Here we actually use the test data and make the prediction by passing the X_test dataset to the 
# model.
y_resampled_pred = lr_resampled_model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [66]:
# Print the balanced_accuracy score of the model 
# Here we use the balanced_accuracy_score() function and pass it the y_test and y_resampled_pred
# datasets as parameters to show us how accurate the prediction was. We can see it was very accurate
# with a 99% accuracy_score.
balanced_accuracy_score(y_test, y_resampled_pred)

0.9936781215845847

In [67]:
# Generate a confusion matrix for the model
# Here again we call on the confusion_matrix() function to show us more in-depth how accurate the model
# was. You can see that 18649 loans were predicted healthy and did not default, 116 loans were
# unhealthy and did not default. 4 loans were predicted healthy and defaulted and 615 were unhealthy
# and defaulted.
confusion_matrix(y_test, y_resampled_pred)

array([[18649,   116],
       [    4,   615]])

In [69]:
# Print the classification report for the model
# Here we have the classification_report_imbalanced() funtion and are able to gain more insight
# into how our model did. Because I imagine we are most concerned with loans that were initially 
# labled healthy and defaulted, seen here as 4.
print(classification_report_imbalanced(y_test, y_resampled_pred))

                   pre       rec       spe        f1       geo       iba       sup

          0       1.00      0.99      0.99      1.00      0.99      0.99     18765
          1       0.84      0.99      0.99      0.91      0.99      0.99       619

avg / total       0.99      0.99      0.99      0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** With the sklearn module we are able to call upon the classification_report() function and get a pretty good idea of how the Logistic Regression algorithim predicted the oversampled data. Here we see the values that the model observed as 0 or healthy loans, 100% of the time the model predicted correctly, and all the values that the model observed as 1 high-risk loans, 84% of the time it predicted them correctly. We can also see under the recall column, 99% of the loans that were healthy did not default and 99% of the loans that were high risk did default. So based on these numbers, I would day that through oversampling the training dataset, we were able to make the model very accurate.